ufs_ffu.c

```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/* Copyright (C) 2019 Western Digital Corporation or its affiliates */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <errno.h>
+#include <asm-generic/int-ll64.h>
+
+#include "ufs.h"
+#include "ufs_cmds.h"
+#include "options.h"
+#include "ioctl.h"
+#include "ufs_ffu.h"
+#include "scsi_bsg_util.h"
+
+#define DEVICE_VERSION_OFFSET 0x1E
+#define FFU_STATUS_ATTR 0x14
+
+enum ffu_status_type {
+	NO_INFORMATION,
+	SUCCESSFUL_MICROCODE_UPDATE,
+	MICROCODE_CORRUPTION_ERROR,
+	INTERNAL_ERROR,
+	MICROCODE_VERSION_MISMATCH,
+	GENERAL_ERROR = 0xFF
+};
+
+
+extern int do_query_rq(int fd, struct ufs_bsg_request *bsg_req,
+		struct ufs_bsg_reply *bsg_rsp, __u8 query_req_func,
+		__u8 opcode, __u8 idn, __u8 index, __u8 sel,
+		__u16 req_buf_len, __u16 res_buf_len, __u8 *data_buf);
+extern struct desc_field_offset device_desc_field_name[];
+
+/* Get sense key string or NULL if not available */
```

Comment:
this can be more economically done using a static array, or even better stringify the enum

Reply:
but this is really just a nitpic, the rest of the code looks fine, so better we'll fix it in some other time, in a different patch